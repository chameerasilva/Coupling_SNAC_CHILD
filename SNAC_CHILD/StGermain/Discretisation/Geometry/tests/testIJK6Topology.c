/*~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
**
** Copyright (C), 2003, Victorian Partnership for Advanced Computing (VPAC) Ltd, 110 Victoria Street, Melbourne, 3053, Australia.
**
** Authors:
**	Stevan M. Quenette, Senior Software Engineer, VPAC. (steve@vpac.org)
**	Patrick D. Sunter, Software Engineer, VPAC. (pds@vpac.org)
**	Luke J. Hodkinson, Computational Engineer, VPAC. (lhodkins@vpac.org)
**	Siew-Ching Tan, Software Engineer, VPAC. (siew@vpac.org)
**	Alan H. Lo, Computational Engineer, VPAC. (alan@vpac.org)
**	Raquibul Hassan, Computational Engineer, VPAC. (raq@vpac.org)
**
**  This library is free software; you can redistribute it and/or
**  modify it under the terms of the GNU Lesser General Public
**  License as published by the Free Software Foundation; either
**  version 2.1 of the License, or (at your option) any later version.
**
**  This library is distributed in the hope that it will be useful,
**  but WITHOUT ANY WARRANTY; without even the implied warranty of
**  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
**  Lesser General Public License for more details.
**
**  You should have received a copy of the GNU Lesser General Public
**  License along with this library; if not, write to the Free Software
**  Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
**
** $Id: testIJK6Topology.c 3462 2006-02-19 06:53:24Z WalterLandry $
**
**~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~*/

#include <mpi.h>
#include "Base/Base.h"

#include "Discretisation/Geometry/Geometry.h"

#include <stdio.h>


void Test_PrintNeighbours( IJK6Topology* topology, Index i, Index j, Index k, Bool dynamicSizes ) {
	Index elNum;
	NeighbourIndex nbrCount;
	NeighbourIndex* elNbrs;
	char* dynStr;

	topology->dynamicSizes = dynamicSizes;
	if ( dynamicSizes ) {
		dynStr = "on";
	} 
	else {
		dynStr = "off";
	}
	
	IJK_3DTo1D_3( topology, i, j, k, &elNum );
	nbrCount = Topology_NeighbourCount( topology, elNum );
	elNbrs = Memory_Alloc_Array( NeighbourIndex, nbrCount, "elNbrs" );
	Topology_BuildNeighbours( topology, elNum, elNbrs );
	printf( "\tElNbrs[0-%d] = {", nbrCount-1 );
	for ( i=0; i < nbrCount; i++ ) {
		printf( "%d", elNbrs[i] );
		if ( elNbrs[i] == Topology_Invalid( topology ) ) {
			printf( "(Inv)");
		}
		printf( ", ");
	}
	printf( "}\n" );
	Memory_Free( elNbrs );
}	


void Test_TestTopologyOfSize( Dictionary* dictionary, Index sizeI, Index sizeJ, Index sizeK ) {	
	IJK6Topology*	topology;
	Index		i,j,k;
	Index elNum;

	Dictionary_Set( dictionary, "meshSizeI", Dictionary_Entry_Value_FromUnsignedInt( sizeI ) );
	Dictionary_Set( dictionary, "meshSizeJ", Dictionary_Entry_Value_FromUnsignedInt( sizeJ ) );
	Dictionary_Set( dictionary, "meshSizeK", Dictionary_Entry_Value_FromUnsignedInt( sizeK ) );

	printf( "*** Construction test ***\n" );
	topology = IJK6Topology_New( "topology", dictionary );
	printf( "topology->size[0-2] = { " );
	for ( i=0; i < 3; i++ ) {
		printf( "%d, ", topology->size[i] );
	}
	printf( "}\n\n" );

	printf( "*** Topology_NeighbourCount() and BuildNeighbour() tests. ***\n\n" );
	for ( k=0; k < sizeK; k++ ) {
		for ( j=0; j < sizeJ; j++ ) {
			for ( i=0; i < sizeI; i++ ) {
				IJK_3DTo1D_3( topology, i, j, k, &elNum );
				printf( "Reference object, loc (%d,%d,%d) (%d)\n", 
					i, j, k, elNum );
				printf( "Dynamic:\n" );
				Test_PrintNeighbours( topology, i, j, k, True );
				printf( "Non-Dynamic\n" );
				Test_PrintNeighbours( topology, i, j, k, False );
				printf( "Non-Dynamic, Periodic\n" );
				topology->isPeriodic[I_AXIS] = True;
				topology->isPeriodic[J_AXIS] = True;
				topology->isPeriodic[K_AXIS] = True;
				Test_PrintNeighbours( topology, i, j, k, False );
				topology->isPeriodic[I_AXIS] = False;
				topology->isPeriodic[J_AXIS] = False;
				topology->isPeriodic[K_AXIS] = False;
				printf( "\n" );
			}
		}	
	}

	Stg_Class_Delete( topology );
}	


int main( int argc, char* argv[] ) {
	Dictionary*	dictionary;
	int		rank;
	int		numProcessors;
	MPI_Comm	CommWorld;
	
	/* Initialise MPI, get world info */
	MPI_Init( &argc, &argv );
	MPI_Comm_dup( MPI_COMM_WORLD, &CommWorld );
	MPI_Comm_size( CommWorld, &numProcessors );
	MPI_Comm_rank( CommWorld, &rank );
	
	Base_Init( &argc, &argv );
	DiscretisationGeometry_Init( &argc, &argv );
	MPI_Barrier( CommWorld ); /* Ensures copyright info always come first in output */
	
	dictionary = Dictionary_New();
	Dictionary_Add( dictionary, "meshSizeI", Dictionary_Entry_Value_FromUnsignedInt( 1 ) );
	Dictionary_Add( dictionary, "meshSizeJ", Dictionary_Entry_Value_FromUnsignedInt( 1 ) );
	Dictionary_Add( dictionary, "meshSizeK", Dictionary_Entry_Value_FromUnsignedInt( 1 ) );

	printf( "+++ 1D Tests +++\n\n" );
	Test_TestTopologyOfSize( dictionary, 3, 1, 1 );
	printf( "+++ 2D Tests +++\n\n" );
	Test_TestTopologyOfSize( dictionary, 3, 3, 1 );
	printf( "+++ 3D Tests +++\n\n" );
	Test_TestTopologyOfSize( dictionary, 3, 3, 3 );

	Stg_Class_Delete( dictionary );
		
	DiscretisationGeometry_Finalise();
	Base_Finalise();
	MPI_Finalize();
	
	return 0;
}	
