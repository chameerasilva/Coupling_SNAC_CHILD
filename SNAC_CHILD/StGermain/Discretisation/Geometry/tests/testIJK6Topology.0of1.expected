StGermain Framework. Copyright (C) 2003-2005 VPAC.
+++ 1D Tests +++

*** Construction test ***
topology->size[0-2] = { 3, 1, 1, }

*** Topology_NeighbourCount() and BuildNeighbour() tests. ***

Reference object, loc (0,0,0) (0)
Dynamic:
	ElNbrs[0-0] = {1, }
Non-Dynamic
	ElNbrs[0-1] = {1, 3(Inv), }
Non-Dynamic, Periodic
	ElNbrs[0-1] = {1, 2, }

Reference object, loc (1,0,0) (1)
Dynamic:
	ElNbrs[0-1] = {2, 0, }
Non-Dynamic
	ElNbrs[0-1] = {2, 0, }
Non-Dynamic, Periodic
	ElNbrs[0-1] = {2, 0, }

Reference object, loc (2,0,0) (2)
Dynamic:
	ElNbrs[0-0] = {1, }
Non-Dynamic
	ElNbrs[0-1] = {3(Inv), 1, }
Non-Dynamic, Periodic
	ElNbrs[0-1] = {0, 1, }

+++ 2D Tests +++

*** Construction test ***
topology->size[0-2] = { 3, 3, 1, }

*** Topology_NeighbourCount() and BuildNeighbour() tests. ***

Reference object, loc (0,0,0) (0)
Dynamic:
	ElNbrs[0-1] = {1, 3, }
Non-Dynamic
	ElNbrs[0-3] = {1, 3, 9(Inv), 9(Inv), }
Non-Dynamic, Periodic
	ElNbrs[0-3] = {1, 3, 2, 6, }

Reference object, loc (1,0,0) (1)
Dynamic:
	ElNbrs[0-2] = {2, 4, 0, }
Non-Dynamic
	ElNbrs[0-3] = {2, 4, 0, 9(Inv), }
Non-Dynamic, Periodic
	ElNbrs[0-3] = {2, 4, 0, 7, }

Reference object, loc (2,0,0) (2)
Dynamic:
	ElNbrs[0-1] = {5, 1, }
Non-Dynamic
	ElNbrs[0-3] = {9(Inv), 5, 1, 9(Inv), }
Non-Dynamic, Periodic
	ElNbrs[0-3] = {0, 5, 1, 8, }

Reference object, loc (0,1,0) (3)
Dynamic:
	ElNbrs[0-2] = {4, 6, 0, }
Non-Dynamic
	ElNbrs[0-3] = {4, 6, 9(Inv), 0, }
Non-Dynamic, Periodic
	ElNbrs[0-3] = {4, 6, 5, 0, }

Reference object, loc (1,1,0) (4)
Dynamic:
	ElNbrs[0-3] = {5, 7, 3, 1, }
Non-Dynamic
	ElNbrs[0-3] = {5, 7, 3, 1, }
Non-Dynamic, Periodic
	ElNbrs[0-3] = {5, 7, 3, 1, }

Reference object, loc (2,1,0) (5)
Dynamic:
	ElNbrs[0-2] = {8, 4, 2, }
Non-Dynamic
	ElNbrs[0-3] = {9(Inv), 8, 4, 2, }
Non-Dynamic, Periodic
	ElNbrs[0-3] = {3, 8, 4, 2, }

Reference object, loc (0,2,0) (6)
Dynamic:
	ElNbrs[0-1] = {7, 3, }
Non-Dynamic
	ElNbrs[0-3] = {7, 9(Inv), 9(Inv), 3, }
Non-Dynamic, Periodic
	ElNbrs[0-3] = {7, 0, 8, 3, }

Reference object, loc (1,2,0) (7)
Dynamic:
	ElNbrs[0-2] = {8, 6, 4, }
Non-Dynamic
	ElNbrs[0-3] = {8, 9(Inv), 6, 4, }
Non-Dynamic, Periodic
	ElNbrs[0-3] = {8, 1, 6, 4, }

Reference object, loc (2,2,0) (8)
Dynamic:
	ElNbrs[0-1] = {7, 5, }
Non-Dynamic
	ElNbrs[0-3] = {9(Inv), 9(Inv), 7, 5, }
Non-Dynamic, Periodic
	ElNbrs[0-3] = {6, 2, 7, 5, }

+++ 3D Tests +++

*** Construction test ***
topology->size[0-2] = { 3, 3, 3, }

*** Topology_NeighbourCount() and BuildNeighbour() tests. ***

Reference object, loc (0,0,0) (0)
Dynamic:
	ElNbrs[0-2] = {1, 3, 9, }
Non-Dynamic
	ElNbrs[0-5] = {1, 3, 9, 27(Inv), 27(Inv), 27(Inv), }
Non-Dynamic, Periodic
	ElNbrs[0-5] = {1, 3, 9, 2, 6, 18, }

Reference object, loc (1,0,0) (1)
Dynamic:
	ElNbrs[0-3] = {2, 4, 10, 0, }
Non-Dynamic
	ElNbrs[0-5] = {2, 4, 10, 0, 27(Inv), 27(Inv), }
Non-Dynamic, Periodic
	ElNbrs[0-5] = {2, 4, 10, 0, 7, 19, }

Reference object, loc (2,0,0) (2)
Dynamic:
	ElNbrs[0-2] = {5, 11, 1, }
Non-Dynamic
	ElNbrs[0-5] = {27(Inv), 5, 11, 1, 27(Inv), 27(Inv), }
Non-Dynamic, Periodic
	ElNbrs[0-5] = {0, 5, 11, 1, 8, 20, }

Reference object, loc (0,1,0) (3)
Dynamic:
	ElNbrs[0-3] = {4, 6, 12, 0, }
Non-Dynamic
	ElNbrs[0-5] = {4, 6, 12, 27(Inv), 0, 27(Inv), }
Non-Dynamic, Periodic
	ElNbrs[0-5] = {4, 6, 12, 5, 0, 21, }

Reference object, loc (1,1,0) (4)
Dynamic:
	ElNbrs[0-4] = {5, 7, 13, 3, 1, }
Non-Dynamic
	ElNbrs[0-5] = {5, 7, 13, 3, 1, 27(Inv), }
Non-Dynamic, Periodic
	ElNbrs[0-5] = {5, 7, 13, 3, 1, 22, }

Reference object, loc (2,1,0) (5)
Dynamic:
	ElNbrs[0-3] = {8, 14, 4, 2, }
Non-Dynamic
	ElNbrs[0-5] = {27(Inv), 8, 14, 4, 2, 27(Inv), }
Non-Dynamic, Periodic
	ElNbrs[0-5] = {3, 8, 14, 4, 2, 23, }

Reference object, loc (0,2,0) (6)
Dynamic:
	ElNbrs[0-2] = {7, 15, 3, }
Non-Dynamic
	ElNbrs[0-5] = {7, 27(Inv), 15, 27(Inv), 3, 27(Inv), }
Non-Dynamic, Periodic
	ElNbrs[0-5] = {7, 0, 15, 8, 3, 24, }

Reference object, loc (1,2,0) (7)
Dynamic:
	ElNbrs[0-3] = {8, 16, 6, 4, }
Non-Dynamic
	ElNbrs[0-5] = {8, 27(Inv), 16, 6, 4, 27(Inv), }
Non-Dynamic, Periodic
	ElNbrs[0-5] = {8, 1, 16, 6, 4, 25, }

Reference object, loc (2,2,0) (8)
Dynamic:
	ElNbrs[0-2] = {17, 7, 5, }
Non-Dynamic
	ElNbrs[0-5] = {27(Inv), 27(Inv), 17, 7, 5, 27(Inv), }
Non-Dynamic, Periodic
	ElNbrs[0-5] = {6, 2, 17, 7, 5, 26, }

Reference object, loc (0,0,1) (9)
Dynamic:
	ElNbrs[0-3] = {10, 12, 18, 0, }
Non-Dynamic
	ElNbrs[0-5] = {10, 12, 18, 27(Inv), 27(Inv), 0, }
Non-Dynamic, Periodic
	ElNbrs[0-5] = {10, 12, 18, 11, 15, 0, }

Reference object, loc (1,0,1) (10)
Dynamic:
	ElNbrs[0-4] = {11, 13, 19, 9, 1, }
Non-Dynamic
	ElNbrs[0-5] = {11, 13, 19, 9, 27(Inv), 1, }
Non-Dynamic, Periodic
	ElNbrs[0-5] = {11, 13, 19, 9, 16, 1, }

Reference object, loc (2,0,1) (11)
Dynamic:
	ElNbrs[0-3] = {14, 20, 10, 2, }
Non-Dynamic
	ElNbrs[0-5] = {27(Inv), 14, 20, 10, 27(Inv), 2, }
Non-Dynamic, Periodic
	ElNbrs[0-5] = {9, 14, 20, 10, 17, 2, }

Reference object, loc (0,1,1) (12)
Dynamic:
	ElNbrs[0-4] = {13, 15, 21, 9, 3, }
Non-Dynamic
	ElNbrs[0-5] = {13, 15, 21, 27(Inv), 9, 3, }
Non-Dynamic, Periodic
	ElNbrs[0-5] = {13, 15, 21, 14, 9, 3, }

Reference object, loc (1,1,1) (13)
Dynamic:
	ElNbrs[0-5] = {14, 16, 22, 12, 10, 4, }
Non-Dynamic
	ElNbrs[0-5] = {14, 16, 22, 12, 10, 4, }
Non-Dynamic, Periodic
	ElNbrs[0-5] = {14, 16, 22, 12, 10, 4, }

Reference object, loc (2,1,1) (14)
Dynamic:
	ElNbrs[0-4] = {17, 23, 13, 11, 5, }
Non-Dynamic
	ElNbrs[0-5] = {27(Inv), 17, 23, 13, 11, 5, }
Non-Dynamic, Periodic
	ElNbrs[0-5] = {12, 17, 23, 13, 11, 5, }

Reference object, loc (0,2,1) (15)
Dynamic:
	ElNbrs[0-3] = {16, 24, 12, 6, }
Non-Dynamic
	ElNbrs[0-5] = {16, 27(Inv), 24, 27(Inv), 12, 6, }
Non-Dynamic, Periodic
	ElNbrs[0-5] = {16, 9, 24, 17, 12, 6, }

Reference object, loc (1,2,1) (16)
Dynamic:
	ElNbrs[0-4] = {17, 25, 15, 13, 7, }
Non-Dynamic
	ElNbrs[0-5] = {17, 27(Inv), 25, 15, 13, 7, }
Non-Dynamic, Periodic
	ElNbrs[0-5] = {17, 10, 25, 15, 13, 7, }

Reference object, loc (2,2,1) (17)
Dynamic:
	ElNbrs[0-3] = {26, 16, 14, 8, }
Non-Dynamic
	ElNbrs[0-5] = {27(Inv), 27(Inv), 26, 16, 14, 8, }
Non-Dynamic, Periodic
	ElNbrs[0-5] = {15, 11, 26, 16, 14, 8, }

Reference object, loc (0,0,2) (18)
Dynamic:
	ElNbrs[0-2] = {19, 21, 9, }
Non-Dynamic
	ElNbrs[0-5] = {19, 21, 27(Inv), 27(Inv), 27(Inv), 9, }
Non-Dynamic, Periodic
	ElNbrs[0-5] = {19, 21, 0, 20, 24, 9, }

Reference object, loc (1,0,2) (19)
Dynamic:
	ElNbrs[0-3] = {20, 22, 18, 10, }
Non-Dynamic
	ElNbrs[0-5] = {20, 22, 27(Inv), 18, 27(Inv), 10, }
Non-Dynamic, Periodic
	ElNbrs[0-5] = {20, 22, 1, 18, 25, 10, }

Reference object, loc (2,0,2) (20)
Dynamic:
	ElNbrs[0-2] = {23, 19, 11, }
Non-Dynamic
	ElNbrs[0-5] = {27(Inv), 23, 27(Inv), 19, 27(Inv), 11, }
Non-Dynamic, Periodic
	ElNbrs[0-5] = {18, 23, 2, 19, 26, 11, }

Reference object, loc (0,1,2) (21)
Dynamic:
	ElNbrs[0-3] = {22, 24, 18, 12, }
Non-Dynamic
	ElNbrs[0-5] = {22, 24, 27(Inv), 27(Inv), 18, 12, }
Non-Dynamic, Periodic
	ElNbrs[0-5] = {22, 24, 3, 23, 18, 12, }

Reference object, loc (1,1,2) (22)
Dynamic:
	ElNbrs[0-4] = {23, 25, 21, 19, 13, }
Non-Dynamic
	ElNbrs[0-5] = {23, 25, 27(Inv), 21, 19, 13, }
Non-Dynamic, Periodic
	ElNbrs[0-5] = {23, 25, 4, 21, 19, 13, }

Reference object, loc (2,1,2) (23)
Dynamic:
	ElNbrs[0-3] = {26, 22, 20, 14, }
Non-Dynamic
	ElNbrs[0-5] = {27(Inv), 26, 27(Inv), 22, 20, 14, }
Non-Dynamic, Periodic
	ElNbrs[0-5] = {21, 26, 5, 22, 20, 14, }

Reference object, loc (0,2,2) (24)
Dynamic:
	ElNbrs[0-2] = {25, 21, 15, }
Non-Dynamic
	ElNbrs[0-5] = {25, 27(Inv), 27(Inv), 27(Inv), 21, 15, }
Non-Dynamic, Periodic
	ElNbrs[0-5] = {25, 18, 6, 26, 21, 15, }

Reference object, loc (1,2,2) (25)
Dynamic:
	ElNbrs[0-3] = {26, 24, 22, 16, }
Non-Dynamic
	ElNbrs[0-5] = {26, 27(Inv), 27(Inv), 24, 22, 16, }
Non-Dynamic, Periodic
	ElNbrs[0-5] = {26, 19, 7, 24, 22, 16, }

Reference object, loc (2,2,2) (26)
Dynamic:
	ElNbrs[0-2] = {25, 23, 17, }
Non-Dynamic
	ElNbrs[0-5] = {27(Inv), 27(Inv), 27(Inv), 25, 23, 17, }
Non-Dynamic, Periodic
	ElNbrs[0-5] = {24, 20, 8, 25, 23, 17, }

